src/components/join-flow/join-flow.jsx

```diff
@@ -13,56 +16,169 @@ const GenderStep = require('./gender-step.jsx');
 const CountryStep = require('./country-step.jsx');
 const EmailStep = require('./email-step.jsx');
 const WelcomeStep = require('./welcome-step.jsx');
+const RegistrationErrorStep = require('./registration-error-step.jsx');
 
-/*
-eslint-disable react/prefer-stateless-function, react/no-unused-prop-types, no-useless-constructor
-*/
 class JoinFlow extends React.Component {
     constructor (props) {
         super(props);
         bindAll(this, [
-            'handleAdvanceStep'
+            'handleAdvanceStep',
+            'handlePrepareToRegister',
+            'handleRegistrationResponse',
+            'handleSubmitRegistration'
         ]);
         this.state = {
             formData: {},
             registrationError: null,
-            step: 0
+            step: 0,
+            waiting: false
         };
     }
-    handleAdvanceStep (formData) {
-        formData = formData || {};
+    handlePrepareToRegister (newFormData) {
+        newFormData = newFormData || {};
+        const newState = {
+            formData: defaults({}, newFormData, this.state.formData)
+        };
+        this.setState(newState, () => {
+            this.handleSubmitRegistration(this.state.formData);
+        });
+    }
+    handleRegistrationResponse (err, body, res) {
+        // example of failing response:
+        // [
+        //   {
+        //     "msg": "This field is required.",
+        //     "errors": {
+        //       "username": ["This field is required."],
+        //       "recaptcha": ["Incorrect, please try again."]
+        //     },
+        //     "success": false
+        //   }
+        // ]
+        this.setState({waiting: false}, () => {
+            let errStr = '';
+            if (!err && res.statusCode === 200) {
```

Comment:
What about when `err` exists? we should handle timeouts, 403, 503, etc. but I don't think those cases are handled. As far as I can tell, we're only handling the case where we get a 200.

Reply:
Yeah, in that case execution will continue after this block ends, and the next step will be to set `registrationError` and include the status code.